File: ClienteGestionAgro/AgropecuarioCliente/Models/ProductoAgricola.cs
using System;
using Newtonsoft.Json;

namespace AgropecuarioCliente.Models
{
    public class ProductoAgricola
    {
        [JsonProperty("id")]
        public string Id { get; set; }

        [JsonProperty("nombre")]
        public string Nombre { get; set; }

        [JsonProperty("hectareasCultivadas")]
        public double HectareasCultivadas { get; set; }

        [JsonProperty("cantidadProducida")]
        public int CantidadProducida { get; set; }

        [JsonProperty("fechaProduccion")]
        public DateTime FechaProduccion { get; set; }

        [JsonProperty("tipoCultivo")]
        public string TipoCultivo { get; set; }

        [JsonProperty("precioVenta")]
        public double PrecioVenta { get; set; }

        [JsonProperty("costoProduccion")]
        public double CostoProduccion { get; set; }

        [JsonProperty("rendimientoPorHa")]
        public double? RendimientoPorHa { get; set; }

        [JsonProperty("temporada")]
        public string Temporada { get; set; } = "Todo el año";

        [JsonProperty("tipoSuelo")]
        public string TipoSuelo { get; set; } = "Franco";

        [JsonProperty("codigoFinca")]
        public string CodigoFinca { get; set; }

        public ProductoAgricola()
        {
            FechaProduccion = DateTime.Now;
        }

        public double CalcularIngresoTotal()
        {
            return CantidadProducida * PrecioVenta;
        }

        public double CalcularRentabilidad()
        {
            if (HectareasCultivadas == 0) return 0;
            double ingresoTotal = CalcularIngresoTotal();
            double costoTotal = CostoProduccion * HectareasCultivadas;
            double utilidadNeta = ingresoTotal - costoTotal;
            return utilidadNeta / HectareasCultivadas;
        }

        public double CalcularMargenGanancia()
        {
            if (CostoProduccion == 0) return 0;
            return ((PrecioVenta - CostoProduccion) / CostoProduccion) * 100;
        }

        public override string ToString()
        {
            return $"{Id} - {Nombre} ({TipoCultivo})";
        }

        // Métodos de validación
        public static class Validations
        {
            public static bool ValidarNombre(string nombre, out string error)
            {
                error = "";
                if (string.IsNullOrWhiteSpace(nombre))
                {
                    error = "El nombre es obligatorio";
                    return false;
                }
                if (nombre.Trim().Length < 2)
                {
                    error = "El nombre debe tener al menos 2 caracteres";
                    return false;
                }
                return true;
            }

            public static bool ValidarHectareas(double hectareas, out string error)
            {
                error = "";
                if (hectareas < 0.1 || hectareas > 10000)
                {
                    error = "Las hectáreas deben estar entre 0.1 y 10,000";
                    return false;
                }
                return true;
            }

            public static bool ValidarCantidad(int cantidad, out string error)
            {
                error = "";
                if (cantidad < 1)
                {
                    error = "La cantidad debe ser mayor a 0";
                    return false;
                }
                return true;
            }

            public static bool ValidarPrecio(double precio, out string error)
            {
                error = "";
                if (precio < 100 || precio > 1000000)
                {
                    error = "El precio debe estar entre $100 y $1,000,000";
                    return false;
                }
                return true;
            }

            public static bool ValidarCosto(double costo, out string error)
            {
                error = "";
                if (costo < 100)
                {
                    error = "El costo debe ser mayor a $100";
                    return false;
                }
                return true;
            }

            public static bool ValidarTipoCultivo(string tipoCultivo, out string error)
            {
                error = "";
                if (string.IsNullOrWhiteSpace(tipoCultivo))
                {
                    error = "El tipo de cultivo es obligatorio";
                    return false;
                }
                return true;
            }
        }
    }
}

File: AgroMicroservicio/src/main/java/co/unibague/agropecuario/rest/model/ProductoAgricola.java
package co.unibague.agropecuario.rest.model;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonProperty;
import jakarta.persistence.*;
import jakarta.validation.constraints.*;
import java.time.LocalDateTime;
import java.util.Objects;

/**
 * Entidad JPA ProductoAgricola - Maestro
 * Universidad de Ibagué - Desarrollo de Aplicaciones Empresariales
 * Tercer Prototipo - Persistencia con MySQL
 */
@Entity
@Table(name = "producto_agricola")
public class ProductoAgricola {

    @Id
    @Column(name = "id", length = 10, nullable = false)
    @Size(max = 10, message = "El ID no puede exceder 10 caracteres")
    private String id;

    @Column(name = "nombre", length = 100, nullable = false)
    @NotBlank(message = "El nombre es obligatorio")
    @Size(min = 2, max = 100, message = "El nombre debe tener entre 2 y 100 caracteres")
    private String nombre;

    @Column(name = "hectareas_cultivadas", nullable = false)
    @NotNull(message = "Las hectáreas cultivadas son obligatorias")
    @DecimalMin(value = "0.1", message = "Las hectáreas deben ser mayor a 0.1")
    @DecimalMax(value = "10000.0", message = "Las hectáreas no pueden exceder 10,000")
    @JsonProperty("hectareasCultivadas")
    private Double hectareasCultivadas;

    @Column(name = "cantidad_producida", nullable = false)
    @NotNull(message = "La cantidad producida es obligatoria")
    @Min(value = 1, message = "La cantidad producida debe ser mayor a 0")
    @Max(value = 1000000, message = "La cantidad producida no puede exceder 1,000,000")
    @JsonProperty("cantidadProducida")
    private Integer cantidadProducida;

    @Column(name = "fecha_produccion")
    @JsonFormat(pattern = "yyyy-MM-dd'T'HH:mm:ss")
    @JsonProperty("fechaProduccion")
    private LocalDateTime fechaProduccion;

    @Column(name = "tipo_cultivo", length = 50, nullable = false)
    @NotBlank(message = "El tipo de cultivo es obligatorio")
    @JsonProperty("tipoCultivo")
    private String tipoCultivo;

    @Column(name = "precio_venta", nullable = false)
    @NotNull(message = "El precio de venta es obligatorio")
    @DecimalMin(value = "100.0", message = "El precio debe ser mayor a $100")
    @DecimalMax(value = "1000000.0", message = "El precio no puede exceder $1,000,000")
    @JsonProperty("precioVenta")
    private Double precioVenta;

    @Column(name = "costo_produccion", nullable = false)
    @NotNull(message = "El costo de producción es obligatorio")
    @DecimalMin(value = "100.0", message = "El costo debe ser mayor a $100")
    @JsonProperty("costoProduccion")
    private Double costoProduccion;

    @Column(name = "rendimiento_por_ha")
    @JsonProperty("rendimientoPorHa")
    private Double rendimientoPorHa;

    @Column(name = "temporada", length = 50)
    private String temporada;

    @Column(name = "tipo_suelo", length = 50)
    @JsonProperty("tipoSuelo")
    private String tipoSuelo;

    @Column(name = "codigo_finca", length = 20)
    @JsonProperty("codigoFinca")
    private String codigoFinca;

    // Constructores
    public ProductoAgricola() {
        this.fechaProduccion = LocalDateTime.now();
        this.temporada = "Todo el año";
        this.tipoSuelo = "Franco";
    }

    public ProductoAgricola(String id, String nombre, Double hectareasCultivadas,
                            Integer cantidadProducida, LocalDateTime fechaProduccion,
                            String tipoCultivo, Double precioVenta, Double costoProduccion) {
        this();
        this.id = id;
        this.nombre = nombre;
        this.hectareasCultivadas = hectareasCultivadas;
        this.cantidadProducida = cantidadProducida;
        this.fechaProduccion = fechaProduccion;
        this.tipoCultivo = tipoCultivo;
        this.precioVenta = precioVenta;
        this.costoProduccion = costoProduccion;
    }

    // Métodos de negocio
    public double calcularRentabilidad() {
        if (hectareasCultivadas == null || hectareasCultivadas == 0) return 0;

        double ingresoTotal = calcularIngresoTotal();
        double costoTotal = costoProduccion * hectareasCultivadas;
        double utilidadNeta = ingresoTotal - costoTotal;

        return utilidadNeta / hectareasCultivadas;
    }

    public double calcularIngresoTotal() {
        if (cantidadProducida == null || precioVenta == null) return 0;
        return cantidadProducida * precioVenta;
    }

    public double calcularMargenGanancia() {
        if (costoProduccion == null || costoProduccion == 0) return 0;
        return ((precioVenta - costoProduccion) / costoProduccion) * 100;
    }

    // Getters y Setters
    public String getId() { return id; }
    public void setId(String id) { this.id = id; }

    public String getNombre() { return nombre; }
    public void setNombre(String nombre) { this.nombre = nombre; }

    public Double getHectareasCultivadas() { return hectareasCultivadas; }
    public void setHectareasCultivadas(Double hectareasCultivadas) { this.hectareasCultivadas = hectareasCultivadas; }

    public Integer getCantidadProducida() { return cantidadProducida; }
    public void setCantidadProducida(Integer cantidadProducida) { this.cantidadProducida = cantidadProducida; }

    public LocalDateTime getFechaProduccion() { return fechaProduccion; }
    public void setFechaProduccion(LocalDateTime fechaProduccion) { this.fechaProduccion = fechaProduccion; }

    public String getTipoCultivo() { return tipoCultivo; }
    public void setTipoCultivo(String tipoCultivo) { this.tipoCultivo = tipoCultivo; }

    public Double getPrecioVenta() { return precioVenta; }
    public void setPrecioVenta(Double precioVenta) { this.precioVenta = precioVenta; }

    public Double getCostoProduccion() { return costoProduccion; }
    public void setCostoProduccion(Double costoProduccion) { this.costoProduccion = costoProduccion; }

    public Double getRendimientoPorHa() { return rendimientoPorHa; }
    public void setRendimientoPorHa(Double rendimientoPorHa) { this.rendimientoPorHa = rendimientoPorHa; }

    public String getTemporada() { return temporada; }
    public void setTemporada(String temporada) { this.temporada = temporada; }

    public String getTipoSuelo() { return tipoSuelo; }
    public void setTipoSuelo(String tipoSuelo) { this.tipoSuelo = tipoSuelo; }

    public String getCodigoFinca() { return codigoFinca; }
    public void setCodigoFinca(String codigoFinca) { this.codigoFinca = codigoFinca; }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        ProductoAgricola that = (ProductoAgricola) obj;
        return Objects.equals(id, that.id);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }

    @Override
    public String toString() {
        return String.format("ProductoAgricola{id='%s', nombre='%s', tipo='%s', hectareas=%.2f}",
                id, nombre, tipoCultivo, hectareasCultivadas);
    }
}